### DISCUSSION TITLE
WGSL

### DISCUSSION THREAD

**Discussion by unicomp21** [stegu/psrdnoise#3] [Ideas]
Has anyone ported to wgsl yet?

**Comment by unicomp21** [comment]
WebGPU is about to go live in March I think?

**Comment by stegu** [comment]
Okay, I just read through the spec for WGSL. It feels a bit rough and
unfinished for a March release, and I still find errors in it, but hey,
this is W3C, and I've seen worse. The syntax reads like a throwback to me,
with elements of disassembly, flashes of Fortran loops, a lack of
overloading for user defined functions, finicky greedy parsing that will
have me swearing when I forget space-after-colon, a modulo operator that is
actually a remainder operator, and a few weird details that will take some
getting used to, but it looks human-writable, and I could probably port the
two psrdnoise functions myself.

If I decide to try it, what do I use for testing?

**Comment by unicomp21** [comment]
Install chrome canary per this article ...

https://web.dev/gpu-compute/

**Comment by stegu** [comment]
Hmm. I can't find any concrete code examples not using Node or Rust. Is
there any basic JS-only template for just pushing a full viewport quad in
view coords and render pixels to a canvas with a pair of WGSL vertex and
fragment shaders? WGSL seems easy enough, but Node and Rust are barriers I
don't have the option to work my way through right now.

Den fre 25 feb. 2022 23:14John Davis ***@***.***> skrev:

**Comment by unicomp21** [comment]
https://github.com/austinEng/webgpu-samples

Does this work?

**Comment by stegu** [comment]
Meh. Node. Thanks for looking, though.
I found a vanilla JS version just now. It's from a few years ago and
required some updates and tweaks, and a change from GLSL to WGSL, but
Chrome gave reasonably informative deprecation messages, and I now have a
basic WGSL shader running from plain JS.

The port is underway, but it won't be done in a jiffy. WGSL is about as
awkward to use as SPIR-V assembly language. Which is natural, of course,
since that is what it's supposed to map to in a bijective and "trivial"
fashion.

I'll post an update when/if anything works out. Right now I'm just
rendering a grid of dots, but at least it's the 2D simplex grid.

**Comment by stegu** [comment]
An incomplete WGSL port of the 2D noise is in the src/ folder. No derivatives yet.
I think WGSL blew it with the modulo operator not really being modulo, the modf() function returning trunc() rather than floor() in an undocumented fashion this close to spec finalization, and the lack of a built-in mod() function, but otherwise the port was relatively painless. Lots of typing, but I'll live.

**Comment by stegu** [comment]
Port of 3D noise added, still without the derivatives. This will have to do for today. Hopefully I can find the time to do the gradient-computing versions soon. It's really just a matter of deciding on the field names for the return struct, but names are important. [n, g] or [noise, gradient]? I will sleep on that.
We're looking at eight functions each for 2D and 3D to implement all combinations of (p?sr?d?noise(2|3)), but seeing how WGSL is all about performance, I decided it's better not to depend on a (reportedly) shaky dead code removal in the compiler, and the lack of overloading means we are facing the return of the "glVertex3fv()" explicit decoration debacle from plain old C no matter what.
I do understand the reasons for WGSL looking like it does, but WGSL will definitely not be my preferred shading language for experiments. I think GLSL/HLSL are at the right level to think and program at the same time, in the manner that I have done software shader programming since the early 1990's (for better or worse, but I like it).
WGSL is just too. much. bloody. typing. Pun totally intended. I have written &lt;f32&gt; more times than I care to count today.

**Comment by unicomp21** [comment]
Thanks @stegu !

**Comment by stegu** [comment]
You're welcome! I enjoyed it, and I needed a reason to learn WGSL.
